### ISSUE TITLE
draco-corrcal pipeline

### ISSUE THREAD

**Issue by piyanatk** [UKZN-Astronomy/hirax_corrcal#12]
We want to develop a pipeline that use `draco`, `corrcal`, and in-house codes to simulate, calibrate, calculate power spectrum, and perform relevant analysis and visualisation.

The pipeline will consist of Python modules and a Jupyter notebook. The latter will be primarily for analysis and visualisation. This pipeline will serve as our development platform for #7, and will fulfil #3, #5, and #10. 

For convenient, we will develop and test this pipeline with a simple set of simulations, but the pipeline will be scalable:

### Simulation Components

- [ ] 2x2 array
- [x] Somewhat realistic beam, e.g. airy beam
- [x] A few point sources
- [x] A few time
- [x] A few frequency
- [ ] 21 cm

### Analysis Components

- [ ] Functions for applying and visualisation of gain solutions
- [ ] Functions for analysis and visualisation of visibility
- [ ] Functions for power spectrum calculation and visualisation

@Zahra-Kader @dcrichton Are you on board? Any suggestions or comments?

**Comment by Zahra-Kader** [comment]
Yes, I'm on board. I will get started on this today and list any questions/concerns that I have.

**Comment by dcrichton** [comment]
Sounds good to me. Although, given the way draco works with caput-pipeline, a jupyter notebook might be a bit annoying for some aspects.

**Comment by piyanatk** [comment]
@dcrichton Can you elaborate? My thought is that this notebook would not just be a pipeline that simply run the codes. We will also use it for documentation, analysis and visualisation. If you mean that we might have to execute Python scripts outside of the notebook and read in thee data, that should be easy to do with magic commands.

**Comment by dcrichton** [comment]
Yeah exactly. It will depend on some configuration files, external script runs and generated outputs which means it won't be super self-contained in a notebook. I think it's fine to use terminal calls or magics in the notebook for this but you'd just want to make sure all the additional files are kept in sync or just have the notebook generate the configs too.

**Comment by piyanatk** [comment]
@Zahra-Kader I think you may want to a new directory in the repo for the notebook, associated scripts, and all input and output files that they produce. You may not need to push all input/output files to the repo, and it might not be possible to do anyway as GitHub has 100MB file limit. It would bee nice still to keep track of the files in the notebooks, maybe as comments.

It might also be handy to try other IPython magics. I think [`%%writefile`](https://ipython.readthedocs.io/en/stable/interactive/magics.html#cellmagic-writefile) is a particularly useful one, so you develop the content of your script as a cell in a notebook, save it out into a file, and in the next cell you can run it with `%run` or `%%bash` or something else. This page lists all of the magic commands https://ipython.readthedocs.io/en/stable/interactive/magics.html.

**Comment by piyanatk** [comment]
I updated the first post to re-scope and include points from today discussion

**Comment by Zahra-Kader** [comment]
I uploaded some details of the working pipeline. So far it can be used to generate corrected visibility timestreams after computing recovered gains with CorrCal. The code currently is not set up to do power spectrum estimation. I need sidereal streams to do that. I still need to show the results of this in a notebook.

**Comment by piyanatk** [comment]
@Zahra-Kader Great! I see you put the gain fitting function in the Python file. That's a good thing. Keep it this way while you are still developing the codes. It will become clear if it'd better be put somewhere else.

For getting the sidereal stream, it is just a matter of simulating visibility time stream from the same LST right? Correct me if I am wrong. Do you foreseen any sort of "utility functions" that we might need?

**Comment by Zahra-Kader** [comment]
With draco, you can get sidereal streams, and from that, the timestreams are generated (see draco/synthesis/stream.py). I currently set up the code to return gain solutions in the shape (nfreq, 2*Ndish, ntime) which is correct for timestream data. I would need to change this if I am to use sidereal streams. I'll look into this more a bit later. I think for now I'll just focus on making sure everything works with the calibrated visibility timestreams.
